USART and USB Bootloader ability added.

ksschwabe opened the pull request:
@davids5: I have updated my pull request for parallel USB and USART bootloading to be in line with the changes that you made to have a single hw_config file.

I don't have a Pixhawk on hand to test the USART bootloading on that board, so please bear this in mind before merging it in - I have put the USART port on USART2 for the Pixhawk. I have tested the USB and USART for the Px4FMU, though.

Here is a quick overview of the changes:

* In the bl.c file I have now changed the bootloader function to accept both the timeout and the bootloader type (either USART or USB) . This option will then set the appropriate functions cin_usb or cin_usart and cout_usb or cout_usart to the function pointers cin and cout. 

* I have changed the #ifdef INTERFACE_USART to be #if INTERFACE_USART. I don't know whether you have a strong opinion on this?

* In order to accommodate the ability to have parallel USB and USART bootload ability, I have had to add some more variables to the main_f4.c file. These are to tell you whether you should try to boot from the USB bootloader or the USART bootloader.

* I have also now given the bootloader an int return type. If it returns 0, it means an app has been uploaded successfully and you can try to boot. If it returns -1, it means that the bootloader has timed-out.

The basic flow of the main program is as follows.
* Enters main.
* If boot signature not set or the USB not connected, try to jump to the app immediately.
* If jumping to the app failed, try the USART bootloader forever.
* If the USB is connected, don't try to boot and try the USB bootloader. If timed-out, try to boot like usual.
* If the USB is disconnected and the boot signature is set, try the USART bootloader.
* If the USB is connected AND the boot signature is set, don't try to boot, then first try the USB bootloader. If you timeout try the USART bootloader. If an app was successfully uploaded by the USB bootloader, it will try to jump_to the_app. If that fails it will then try the USART bootloader.

I have seen that someone has requested the ability to bootload via the USART in issue #33. I am concerned that they might want to bootload via a wireless connection. If this is the case then we would have to add logic to handle bit and byte drops. I don't know whether this is necessary to do at this time. This would slow down the uploading of firmware a bit, so maybe there should be logic so that it only does bit and byte drop checking and compensation if it is bootloading over the USART port. Thoughts?

mhkabir commented:
@ksschwabe Nice! You tested your PR on something like a Discovery board, I presume?

ksschwabe commented:
@mhkabir: No, on a Px4FMU.

magicrub commented:
Thanks @ksschwabe ! Wireless uploading is certainly a plus. I have a very reliable serial->Bluetooth to companion computer link but with wireless it's not common to get 100% error-free. If there is any error at all, what is the behavior? Does it simply fail and stay in the bootloader for a retry? That would be OK.

ksschwabe commented:
@magicrub: The current logic is that if there is an error, the upload will fail and the board will stay in the bootloader waiting for another attempt. You would then have to start the upload again from the PC. Also note that the upload times over the UART port are substantially slower than what you get on the USB port (+/-75s on UART vs +/-15s on USB).

I suspect that you will not be able to get reliable firmware uploads via a wireless link without the bootloader handling bit and byte drops. This logic would still have to be added. I'd like to see what the Px4 repo maintainers have to say about this.

magicrub commented:
@ksschwabe So it's safe to assume there's some sort of error detection somehow? CSum? Is thta on a packet basis or at the end of the upload? Is there any notification to the PC that the upload has failed mid-upload due to bit/byte errors so it doesn't just keep pushing data? That is, if it fails is it easy for the PC to restart in a timely manner. I'm picturing an upload with repeated retries until it gets it right.

mday299 commented:
+1 If this works we need it here!

magicrub commented:
btw, this is a huge feature that will allow my fleet's companion computers to get the firmware and then deploy it. Very huge. The more robust we can make this, the better!

LorenzMeier commented:
Nice! @davids5 It would be great to add per-sector validation and rev the BL version. That way occasional byte drops are a no-brainer. Could you add the integration of this and the addition of sector-level validation to your list? Its something we want in general and which we just never got around to it.

ksschwabe commented:
@magicrub: At the moment, the error detection is very basic but it is good enough when working with solid copper cables. :-) 

The PC and the bootloader work serially. When the PC sends a packet to write to flash it waits for an OK from the bootloader before sending the next packet. If it receives a "not ok" or FAILED response, then it terminates the upload. Once the bootloader has started a bootload sequence (i.e. received the first valid command from the PC), it will remain in the booloader until it has successfully uploaded a firmware image or the power to the board is disconnected. This means that if the upload fails and you just restart the upload from the PC, the bootloader should just restart with erasing the flash and trying to upload the firmware again.

At the end of the firmware upload, there is a checksum/CRC that is calculated by the bootloader from what has been programmed and it is compared with a CRC for the firmware file that is sent to it from the PC. If they correspond then the bootloader exits and jumps to the main program. If they don't then an error/FAILED response is sent back to the PC and the board stays in the bootloader.

mikeclement commented:
Another +1 for our team!

davids5 commented:
@ksschwabe 

Thank you for this work.

Assuming run_bl = (Valid RTC request to boot || !Vaid(word[0])

When run_bl is true:

Originally what I had envisioned doing was to have both drivers (if they will fit) and in the case of a device that has VBUS detection, and VBUS was active, then attempt the boot from the USB CDC, if VBUS was not active use the serial port.

Then I encountered a device without VBUS detection wired up. So I thould if we have a non blocking getchar on all interfaces we would could just scan the inputs a look for the one that has incoming data that matches the BL protocol then select it (with timeout) and attempt to but from it. 

The downside is that unless the reboot to bootloader command is issued (setting the RTC signature) or VBUS is detectable and active, Serial bootloading would never run on a good load of a bad image. (one not capable of receiving the "reboot to bootloader command") The only recovery would be to have a pin we could sample to force bootloading or check the serial input for 100ms for data as to not delay booting. 

What are your thoughts on this?

davids5 commented:
For reverance I have attached some UML of what the current protocol over serial does

![image](https://cloud.githubusercontent.com/assets/1945821/9613193/9b87cd70-5085-11e5-8ed9-25cf3071eb6a.png)
![image](https://cloud.githubusercontent.com/assets/1945821/9613224/cb0fe988-5085-11e5-8565-1cd4408b7492.png)
![image](https://cloud.githubusercontent.com/assets/1945821/9613234/d5b55742-5085-11e5-8f77-5fa01650d8cf.png)

.

magicrub commented:
Is there room to implement both: force USB if VBus detected and connected, else poll both for data for 100ms (or whatever small timeout).

ksschwabe commented:
@davids5 : I played around with the bootloader a bit this afternoon, and it is certainly possible to scan both the USB and the USART for an incomming signal and if it has a proper "insync" message then continue to communicate on that port. I was unable to get the query time down to 100ms, though. I needed to have it at least 300ms to have it work some of the time, and at least 500ms to have it work most of the time. I would certainly rather have it jump to the main app as soon as possible and don't want to wait 500ms every time I boot.

Perhaps we can have some logic that if there is a "force_bl_pin" set, then only force the bootloader to run then, and scan the USB and/or USART (depending on what someone has compiled in). If you don't have a "force_bl_pin", then you will have to wait for some timeout that we can configure in the hw_config, or in the RTC backup registers.

If VBUS is detected, then definitely wait for the bootloader to timeout as before. For the boards without VBUS, they will have to have a "sample pin" set to force the bootloader, or if they don't have such a pin set, then their bootloader will have to go through a bootloader query timeout every time.

Out of interest, which boards do not have the VBUS detect connected?

davids5 commented:
@ksschwabe 

Thank you for looking into the timing. I agree that the large wait is too long. We could try to use break detection on the the serial port. That way 1) we do not need a special pin, 2) we put the burden on the bootloader side of a the wait. Given the baud rate, the detection time should be really short.

Would you like to try that?

> which boards do not have the VBUS detect connected

The PX4 ESC 1.6 HW - but it uses a UAVCAN bootloader

magicrub commented:
If we're going through all this trouble for just the special case of PX4ESC 1.6, can we simply request a change for v1.7? They've only made a small run of them and I'm sure they could make a minor change like this for the next run. btw, I have a PX4ESC v1.6 on-hand. And like you've said, it's CAN anyway so maybe that exempts it? That's a different bootloader entirely and maybe it can tolerate a 500ms boot delay?

davids5 commented:
@magicrub 

Yes Tom you are correct it currently does exempt it. For now, we are just investigating the "Any port" bootloader as a general mechanism. 

A desirable long term goal would be to have USB, Serial, **UAV**CAN, and X. The limiting factors are the size of the composite bootloader image(s) and the Flash size and granularity on the different SoC.

>  special case of PX4ESC 1.6, can we simply request a change for v1.7? 

Unfortunately the the VBUS detect pin is in use in the motor control drive, and can not be reallocated.

I would like to collect some input here, if we step back and look at the bootloading process: 

If the device has USB and VBUS then - Why do we need serial?
  Prior builds where compile time selection.

ksschwabe commented:
@davids5
The PX4ESC doesn't use this bootloader does it? I have seen in issue PX4/px4esc#2, that there was a long discussion relating to creating a bootloader for the ESC and on PX4/Firmware#1161 Lorenz thanked you for solving the issue, but I haven't seen where this UAVCAN bootloader code is. Am I missing something?

>We could try to use break detection on the the serial port
Unfortunately I know nothing about break detection, but if you can point me in the right direction, I'll gladly give it a go.

ksschwabe commented:
@davids5
>If the device has USB and VBUS then - Why do we need serial?
Prior builds where compile time selection.

I don't know about others, but the main push for me to have the dual bootloader ability was for when I had built the FMU into a frame and couldn't necessarily easily access the USB port again. Then I could always plug in my FTDI cable into my exposed UART port.

I think others would have similar reasons and possibly the potential ability to bootload wirelessly is appealing to some, although we would definitely need to do sector validation and resending of packets if necessary then.

magicrub commented:
I second the wireless programming - huge perk. "Wireless" is a relative term though. Here's a use case for serial via "wireless":
@mday299  who chimed in earlier on here and is one of the guys working on this project: https://www.linkedin.com/pulse/nps-arsenl-successfully-flies-50-uav-swarm-live-fly-field-chung
is managing a fleet of 50+ planes. If you have one tiny update then it becomes a huge logistical chore to deploy it. Deploying via WiFi to companion computer (FTP script or something?) which can verify it then have it upload via serial.

I have a similar use case of having to deploy FW to a fleet and walking up with a USB cable is not preferred.

davids5 commented:
@ksschwabe 

Thank you for the feedback. That is a compelling reason.

> I haven't seen where this UAVCAN bootloader code is. Am I missing something?

The UAVCAN bootloader is built on https://github.com/PX4/Firmware/tree/nuttx_next, it uses nuttx in a baremetal envronment.

> Unfortunately I know nothing about break detection, but if you can point me in the right direction, I'll gladly give it a try.

## The basic concept:
http://ltxfaq.custhelp.com/app/answers/detail/a_id/736/~/what-is-a-serial-break%3F

So on the TTL (3.3V) signal side of the world you would normally see a high level during idle. So if one enables a pull up on the IO pin (and/or the serial input), when a break is sent from the host the pin will go low for longer than a character time at the given board rate.

Have a look at the STM32 Reference manuals. For the STM32F427 see RM0090 

![image](https://cloud.githubusercontent.com/assets/1945821/9640305/b13f7772-514c-11e5-9c3c-7ffe93f62f5f.png)


![image](https://cloud.githubusercontent.com/assets/1945821/9640254/7995ae68-514c-11e5-9c78-d6721b8113fa.png)

Let me know if you need anything else. You can skype me at david_s5.

mhkabir commented:
@ksschwabe Any updates on this? ;)

ksschwabe commented:
@mhkabir : I haven't forgotten about this. I just haven't had much time lately to debug it properly. And I am away from tomorrow until the 4th Oct. Thereafter I can have a proper look at it again.

The current idea is that in addition to having a pin or set of pins that one can read to trigger the bootloader, one can also detect it directly from the the UART port. The problem is that different devices, radios etc. that I have tried seem to send a LIN break slightly differently. On one wireless module that I have, if I "activate" the break condition in the python uploader, it does what I expect and sends a start bit and then is low for 10 bits before sending another start bit and continues. When I use my FTDI cable however, the line is just pulled low and remains low until I release the break condition and then the line goes high again.

It seems that the UART on the STM32F4 only recognises that it is a break if the line started at high at some point, then dropped low for more than 10 bits AND then came back high again. This means to get this to work correctly I need to try and synchronise the break event from the computer and the turning on of the Px4. This is nearly impossible to to especially if I want to keep the "checking" done by the bootloader to be a as short as possible so that it still boots quickly.

What I think I will have to do is reconfigure the UART to be a timer pin and then count the low bits for a two cycles and if I detect the line as being low the entire time or if I received only two high bits, then I can assume that I have received a break event and I can activate the bootloader. I still need to try this though.

If you have any ideas on what else I can try or should try let me know! :-)

ksschwabe commented on main_f4.c:
```diff
@@ -331,6 +416,19 @@ main(void)
 	board_init();
 
+	/* configure the clock for bootloader activity */
+	rcc_clock_setup_hse_3v3(&clock_setup);
```
I have had to move the clock setup earlier in the code, so that I can setup the timer correctly to sample the USART RX pin in order to determine if a break is being received.

davids5 replied:
This is a good point.  By moving this code here the application may be called with the System Clock, uart, a timer and USB configured. This puts the onus on the App to reset and initialize the HW. This could be a problem. So jump_to_app once it determines it will make the jump should revert the SOIC to the state it would have been in like the original code. 

cfin is called on both uart and usb - but they do not use the RCC to reset the peripherals.

The timer is disabled but the RCC is not reset - this could be added.

System clock is not reset back to power on default. - this should be done.

ksschwabe commented on main_f4.c:
```diff
@@ -331,6 +416,19 @@ main(void)
+
+	/* start the interface */
+#if INTERFACE_USART
+	cinit(BOARD_INTERFACE_CONFIG_USART, USART);
```
The interfaces also now have to be setup earlier in the code, so that if the USART is compiled in, I can check if a LIN break was received.

davids5 replied:
Is the break detect unreliable? If it is just sampling the IO pin would help decomplicate reverting of the SoC to the reset state. If we do not have to init the uart, but just sample the pins. 

To further decouple things. We could use the SysTic timer to time the RX pin instead of timer 2. It is running at 1ms but can be sampled - using code from https://github.com/PX4/Firmware/blob/nuttx_next/src/drivers/bootloaders/include/timer.h#L341-L451 and https://github.com/PX4/Firmware/blob/nuttx_next/src/drivers/bootloaders/src/arch/stm32/drivers/can/driver.c#L154-L191

We can time the low time of the RX pin.

ksschwabe replied:
It depends on what you mean by reliable. If when the USART is initialised and the RX line is high and then it goes low for the break and then high again (as specified in the reference manual) then it is reliable. (But for it to be detected it has to have been high before AND after the break for the break flag to be set in the register).

Unfortunately it is just about impossible to have the line be high on startup and then pull it low for the break by the time the booloader sequence checks the RX pin. That is why I have put in the pin sampling stuff. But I see your point: if we basically never detect the break, but rather rely almost entirely on the sampling, then why not just do the sampling? I'll add this in shortly.

ksschwabe commented on main_f4.c:
```diff
@@ -175,13 +182,78 @@ board_test_force_pin()
+	rcc_peripheral_enable_clock(&RCC_APB1ENR, RCC_APB1ENR_TIM2EN);
+
+	//Time base configuration
+	timer_set_prescaler(TIM2, 0); // Timer ticks at base clock speed (2 x APBPB1 = 84MHz)
```
Currently I have hardcoded in the use of TIM2. Do we want to put this to be configurable in hw_config.h?

davids5 replied:
I do not think that would be necessary - but as suggested above if we can use SysTic we would have less things that need to be reverted before the jump to the app.

ksschwabe replied:
I have now changed it to use the SysTick timer instead.

ksschwabe commented on main_f4.c:
```diff
@@ -377,7 +475,15 @@ main(void)
+	 * Check if the force-bootloader pins are strapped; if strapped,
+	 * don't try booting.
+	 */
+	if (board_test_force_pin()) {
```
Previously, the bootloader would not have been activated on boot if the pins were strapped. It would only stay in the bootloader after a bootloader timeout.

I have now put this check here so if the pins are strapped it will trigger the bootlaoder. Is this the functionality that you want?

davids5 replied:
I this this ensures you can replace a good load of bad code. The previous code could have jumped to bad code that was loaded and had a good CRC. - @ksschwabe This is a very good find!

ksschwabe commented on main_f4.c:
```diff
@@ -431,6 +550,12 @@ main(void)
 
+#if INTERFACE_USART
+		/* if the USART port RX line is still receiving a break, just loop back */
+		if (board_test_usart_receiving_break())
```
If the USART port is still receiving a break, loop back like if the force-bl pins were strapped. Do you want this here?

davids5 replied:
This makes sense - but it may also need to wait out the break or handle the data of the break in the bootloader code by ignoring the data with a timeout.

ksschwabe commented:
@davids5 : The bootloader can now be triggered by a LIN break being detected on the USART RX line. There is no guarantee that a break would be detected, by the dedicated USART hardware on the STM32F4, since it requires a line to have been high, and then have dropped to low for 10 bits and then come back up to high. It is just about impossible to be able to do this synchronised with the px_uploader program, so I have ended up sampling the RX pin and "counting" the bits at double the bit rate frequency, i.e. I should get two samples of each bit.

I did did this sampling and timing in the function board_test_usart_receiving_break() using TIM2 (perhaps make this a configurable timer in hw_config.) I would have liked to have used a timer with a capture compare pin to count the bits automatically, but not all of the USART ports on the F4 have timers associated with the same pins (e.g. USART2_RX on PD6), so I have had to make use of this busy_wait methodology.

I have coded it so that if the RX line is also held low, the BL will be triggered.

When the bootloader is triggered it "scans" the USART and USB ports (if both have been configured in hw_config.h) for a command. After the first valid command has been received on a port, it then continues to read and write only on that port for the rest of the bootloading sequence. This means that you can trigger the BL by plugging in the USB cable (PA9 is high) and then uploading the firmware via the USART, or trigger the BL by holding the USART RX line low, then plugging in the USB cable and uploading via the USB cable.

I have tested this using a PX4FMUv1 and also on a custom board running an STM32F27 (like the Pixhawk). I have also tested triggering the BL using an FDTI cable and a wireless module sending break commands. The only problem that I have is that the wireless module might not have received power for long enough to have booted itself before the Px4 bootloader had run through and booted the board. It has worked consistently if the wireless module and the board were already powered, and then the board was "reset" such that the wireless module remained powered and the Px4 then received the break from the wireless.

The bootloader is still "unreliable" in terms of bit and byte drops so we can't yet do a proper bootloader over wireless, but I think we push that as a separate pull request where we increase the BL revision to 5.

magicrub commented:
Thanks so much for working on this! Your staged release method of USART now
(v4) and wireless USART later (v5) sounds reasonable. People will no doubt
try it though so it will have to fail elegantly via crc/sum check at the
end and, of course, always recoverable via USB.
On Oct 12, 2015 4:01 AM, "ksschwabe" <notifications@github.com> wrote:

> @davids5 <https://github.com/davids5> : The bootloader can now be
> triggered by a LIN break being detected on the USART RX line. There is no
> guarantee that a break would be detected, by the dedicated USART hardware
> on the STM32F4, since it requires a line to have been high, and then have
> dropped to low for 10 bits and then come back up to high. It is just about
> impossible to be able to do this synchronised with the px_uploader program,
> so I have ended up sampling the RX pin and "counting" the bits at double
> the bit rate frequency, i.e. I should get two samples of each bit.
>
> I did did this sampling and timing in the function
> board_test_usart_receiving_break() using TIM2 (perhaps make this a
> configurable timer in hw_config.) I would have liked to have used a timer
> with a capture compare pin to count the bits automatically, but not all of
> the USART ports on the F4 have timers associated with the same pins (e.g.
> USART2_RX on PD6), so I have had to make use of this busy_wait methodology.
>
> I have coded it so that if the RX line is also held low, the BL will be
> triggered.
>
> When the bootloader is triggered it "scans" the USART and USB ports (if
> both have been configured in hw_config.h) for a command. After the first
> valid command has been received on a port, it then continues to read and
> write only on that port for the rest of the bootloading sequence. This
> means that you can trigger the BL by plugging in the USB cable (PA9 is
> high) and then uploading the firmware via the USART, or trigger the BL by
> holding the USART RX line low, then plugging in the USB cable and uploading
> via the USB cable.
>
> I have tested this using a PX4FMUv1 and also on a custom board running an
> STM32F27 (like the Pixhawk). I have also tested triggering the BL using an
> FDTI cable and a wireless module sending break commands. The only problem
> that I have is that the wireless module might not have received power for
> long enough to have booted itself before the Px4 bootloader had run through
> and booted the board. It has worked consistently if the wireless module and
> the board were already powered, and then the board was "reset" such that
> the wireless module remained powered and the Px4 then received the break
> from the wireless.
>
> The bootloader is still "unreliable" in terms of bit and byte drops so we
> can't yet do a proper bootloader over wireless, but I think we push that as
> a separate pull request where we increase the BL revision to 5.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PX4/Bootloader/pull/36#issuecomment-147365964>.
>

magicrub commented:
Just thinking out loud here in the case that a "reliable transfer" for
worked serial won't fit in the Bootloading program space... I wonder if
there is any "reliable" mode for the wireless serial module that we could
somehow put it into temporarily. I'm picturing a TCP type mode that ensures
reliable transfer in exchange for bandwidth and is handled in the module
itself instead of the Bootloader. Could be a firmware we write for it if
needed but I'm sure someone has done it by now since it's becoming a pretty
popular chip.
On Oct 12, 2015 5:31 AM, "Tom Pittenger" <tom@airphrame.com> wrote:

> Thanks so much for working on this! Your staged release method of USART
> now (v4) and wireless USART later (v5) sounds reasonable. People will no
> doubt try it though so it will have to fail elegantly via crc/sum check at
> the end and, of course, always recoverable via USB.
> On Oct 12, 2015 4:01 AM, "ksschwabe" <notifications@github.com> wrote:
>
>> @davids5 <https://github.com/davids5> : The bootloader can now be
>> triggered by a LIN break being detected on the USART RX line. There is no
>> guarantee that a break would be detected, by the dedicated USART hardware
>> on the STM32F4, since it requires a line to have been high, and then have
>> dropped to low for 10 bits and then come back up to high. It is just about
>> impossible to be able to do this synchronised with the px_uploader program,
>> so I have ended up sampling the RX pin and "counting" the bits at double
>> the bit rate frequency, i.e. I should get two samples of each bit.
>>
>> I did did this sampling and timing in the function
>> board_test_usart_receiving_break() using TIM2 (perhaps make this a
>> configurable timer in hw_config.) I would have liked to have used a timer
>> with a capture compare pin to count the bits automatically, but not all of
>> the USART ports on the F4 have timers associated with the same pins (e.g.
>> USART2_RX on PD6), so I have had to make use of this busy_wait methodology.
>>
>> I have coded it so that if the RX line is also held low, the BL will be
>> triggered.
>>
>> When the bootloader is triggered it "scans" the USART and USB ports (if
>> both have been configured in hw_config.h) for a command. After the first
>> valid command has been received on a port, it then continues to read and
>> write only on that port for the rest of the bootloading sequence. This
>> means that you can trigger the BL by plugging in the USB cable (PA9 is
>> high) and then uploading the firmware via the USART, or trigger the BL by
>> holding the USART RX line low, then plugging in the USB cable and uploading
>> via the USB cable.
>>
>> I have tested this using a PX4FMUv1 and also on a custom board running an
>> STM32F27 (like the Pixhawk). I have also tested triggering the BL using an
>> FDTI cable and a wireless module sending break commands. The only problem
>> that I have is that the wireless module might not have received power for
>> long enough to have booted itself before the Px4 bootloader had run through
>> and booted the board. It has worked consistently if the wireless module and
>> the board were already powered, and then the board was "reset" such that
>> the wireless module remained powered and the Px4 then received the break
>> from the wireless.
>>
>> The bootloader is still "unreliable" in terms of bit and byte drops so we
>> can't yet do a proper bootloader over wireless, but I think we push that as
>> a separate pull request where we increase the BL revision to 5.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/PX4/Bootloader/pull/36#issuecomment-147365964>.
>>
>

ksschwabe commented:
@magicrub : Currently with a wireless upload, you are likely to get the following "Errors" from the px_uploader:
* unexpected <char> instead of INSYNC
* bootloader reports OPERATION FAILED
* unexpected response <char> instead of OK

All of these "errors" are caused by dropped bytes. If they occur you would just have to start the upload again. The board would remain in the bootloader forever, since there is no longer any valid firmware to boot to. You can still recover the board and upload a firmware via either the USB or the USART.

We will have to see what we can do about the size of the BL when it also handles dropped bits and bytes. I think it should be ok, but we are a bit close to the 4kB max size on the PX4IO bootloaders at the moment (currently sitting at 3.6kB).

LorenzMeier commented:
@ksschwabe Do you feel this is ready to go?

magicrub commented:
There was talk of bumping the version number but don't see it in the
commits.

I see that Mission Planner only supports BL ver 2, 3 and 4:
https://github.com/diydrones/MissionPlanner/blob/master/ExtLibs/px4uploader/Uploader.cs#L71
Not sure about QGC or APM Planner or MAVProxy.

-TomP


On Mon, Oct 26, 2015 at 11:58 AM, Lorenz Meier <notifications@github.com>
wrote:

> @ksschwabe <https://github.com/ksschwabe> Do you feel this is ready to go?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PX4/Bootloader/pull/36#issuecomment-151248527>.
>

ksschwabe commented:
@LorenzMeier : This should be good to go. I've used it on the PX4FMU-v1 board for the last 2 weeks and have been using it for the last week on a custom board. I haven't been able to test it on a Pixhawk, though, since I don't have one, so I would appreciate it if someone could test it on a Pixhawk before merging in the commit. 

To activate the bootloader when you don't have the USB cable plugged in, you can continually send "breaks" on the USART or pull the USART RX line to ground on power up. This will trigger the bootlaoder "timeout" sequence. You can then upload via either the USB or the USART.

@magicrub : We were going to bump the version number only once we account for dropped bits and bytes. Since the overall logic of the bl.c file is the same we will leave the version number as if for the moment.

ksschwabe commented:
@LorenzMeier : I added a few line comments to the code in this PR with some questions regarding some of the functionality. Please could you or @davids5 go through them before merging in this PR.

LorenzMeier commented:
Will do. David planned to get this in within 1-2 days, so it should not be sitting any longer here. We also plan to do tests with a ESP8266 based dongle and upgrades via tablet in the next days.

davids5 commented:
@ksschwabe Thank you for your efforts - I will be reviewing this in the AM.

davids5 commented:
@ksschwabe, FYI @LorenzMeier 

I have reviewed the code changes and the only concern I have is the state of the CPU (SoC) at the time we jump to the app. This is explained in detail in the response to the line comments. Please let me know if you think we should add the code to revert the chip.

Also how would you feel about using SysTic and avoiding Tim2?


During testing on FMUV2 hw under linux (on a real not VM machine) I ran the the master build of Firmware and used ```make  px4fmu-v2_default upload``` to test with.

I saw "timed out waiting for erase" a few times and also "ERROR: timeout waiting for data (4 bytes)" once. So I need to test this more before merging.

ksschwabe commented:
@davids5 

I think we should put in the code for reverting the CPU to its power-on state before jumping to the app. I will try to do this soon. Does this involve just resetting all of the RCC registers to their default values or is it more complicated? Do I also have to reset all the pins to being floating inputs? Also, we should probably reset the main system clock as well (set it back to using the HSI).

Re using the SysTick timer: I have already replaced TIM2 with the SysTick timer. I am currently doing the same thing as what I was doing with TIM2, but I think I might change the way I sample to be a bit more intelligent (based on one of your comments).

Re your problems with uploading: Are you using a cable or a wireless device. I haven't had any problems with a physical cable. Both of those errors that you got are owing to a bits or bytes being dropped.

davids5 commented:
@ksschwabe 


Yes the resets should be just after cfin is called in jump to app. 

I would suggest we use a balanced pattern xxxx_init and xxx_deinit for all resources. So board, systic, CPU clock and uart/usb. 

Then the xxxx_deinit function that is called should tear down the HW from the core outward and ultimately just reset the device block and then disable it the RCC.

So for example UART it sets it to disabled as it does now. The toggle the reset bit in the RCC then disable the UART in the RCC.

The order should be
init 
-----
board
PLL
SYSTIC
UART/USB


deinit
-------
USB/UART
SYSTIC
PLL
BOARD

The last 2 on dinit my be swappable.

davids5 commented:
>Re your problems with uploading: Are you using a cable or a wireless device. I haven't had any problems with a physical cable. Both of those errors that you got are owing to a bits or bytes being dropped.

I was testing wired - I really need to insure my test setup is A OK, I will retest once you have the changes in.

I wonder if the 1000 ms delay is prudent for the wait for EOC. At 115200 the character timeas a 86.8 uS The data is sent back to back. So it seems like 2 ms would work (2 to avoid the quantisation error)
Also a inline bool wait_for_eoc(int time) would be a welcome clean up.

ksschwabe commented:
@davids5 

* I have put in the code to revert the CPU to its power-on state. Have a look at my line-comments in this last commit for details and questions. There is now a board_init, board_deinit function and a clock_init and clock_deinit functions. 

* I moved the initialisation of the force_bl pins into board_init function to be consistent with here all of the other peripherals were being set up (previously it was set up in the test_force_bl function).

* I reduced the EOC wait time down to 2ms as requested and added the inline bool wait_for_eoc() function.

* I removed the code for testing if the USART LIN break bit has been set, and now only do a sampling method to determine if the the break has been received.

ksschwabe commented on bl.c:
```diff
@@ -224,13 +259,15 @@ jump_to_app()
 
-	/* and set a specific LED pattern */
-	led_off(LED_ACTIVITY);
-	led_on(LED_BOOTLOADER);
```
I have now removed this setting of the LEDs prior to the jump, since I am deinitialising them in the board_deinit function.

ksschwabe commented on bl.c:
```diff
@@ -224,13 +259,15 @@ jump_to_app()
+	clock_deinit();
+
+	/* deinitialise the board */
+	board_deinit();
```
Here I am performing the clock and board deinitialisations.

ksschwabe commented on cdcacm.c:
```diff
@@ -296,6 +296,19 @@ usb_cfini(void)
+	/* Reset the USB pins to being floating inputs */ 
+	gpio_set_mode(GPIOA, GPIO_MODE_INPUT,GPIO_CNF_INPUT_FLOAT, GPIO8);
+	gpio_clear(GPIOA, GPIO8);
+#endif
```
I have put the deinitialisations of the USB pins here in the cdcacm driver usb_cfini() function, since they are initialised in the usb_cinit function.

ksschwabe commented on main_f4.c:
```diff
@@ -127,12 +134,6 @@ board_test_force_pin()
-	gpio_mode_setup(BOARD_FORCE_BL_PORT, GPIO_MODE_INPUT, BOARD_FORCE_BL_PULL, BOARD_FORCE_BL_PIN_IN);
-	gpio_mode_setup(BOARD_FORCE_BL_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, BOARD_FORCE_BL_PIN_OUT);
-	gpio_set_output_options(BOARD_FORCE_BL_PORT, GPIO_OTYPE_PP, GPIO_OSPEED_100MHZ, BOARD_FORCE_BL_PIN_OUT);
-
```
I moved these initialisations into the board_init function to be consistent with where all of the other peripherals are set.

ksschwabe commented on main_f4.c:
```diff
@@ -205,10 +296,105 @@ board_init(void)
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPFEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPGEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPHEN);
+	rcc_peripheral_disable_clock(&RCC_AHB1ENR, RCC_AHB1ENR_IOPIEN);
```
Should I individually reset each of the RCC peripheral clocks like what I am doing here, or should I just reset the entire AHB1, AHB2, APB1 and APB2 enable registers back to their defaults?

davids5 replied:
It would be good to just set the RCC_AHB1ENR to the default.

ksschwabe commented on main_f4.c:
```diff
@@ -205,10 +296,105 @@ board_init(void)
+
+	/* Reset the CIR register */
+	RCC_CIR = 0x000000;	
+}
```
Please could you check whether I have missed anything here.

Basically I have:
* turned on the HSI oscillator again
* turned off the HSE oscillator, PLL and CSS
* reset the CFGR, PLLSFGR and CIR registers.

davids5 replied:
This looks good to me.

ksschwabe commented on main_f4.c:
```diff
@@ -403,15 +617,22 @@ main(void)
+#if INTERFACE_USB
+	cinit(BOARD_INTERFACE_CONFIG_USB, USB);
+#endif
+
```
I have moved the intialisations of the interfaces back down here.

davids5 commented:
@ksschwabe,  Thank you for all the changes.  FYI @LorenzMeier  

The changes all look really good. I will pull and test this in the AM on a FMUV2 HW with both USB and Battery powered for both USB and serial bootloading.

davids5 commented:
@ksschwabe - this looks Great - Thank you!

davids5 commented:
@ksschwabe 

I found one issue with the wait_for_eoc changes.

The FMU when updating the the pio
Uses https://github.com/PX4/Firmware/blob/master/src/drivers/px4io/px4io_uploader.cpp#L661

This is a SW timeout and it is set CONFIG_BOARD_LOOPSPERMSEC. 

So it uses a 100Ms delay and the original corresponding delay was  1000Ms

This arrangement would accommodate a 10X error in CONFIG_BOARD_LOOPSPERMSEC, which I think we should keep.

davids5 commented:
@ksschwabe Just a second FYI - There was a second timing issue with PROTO_PROG_MULTI	  So I in creased that timeout.

ksschwabe commented:
@davids5 : Thanks for the info! :+1:

magicrub commented:
I don't see in the code where it chooses which USART to configure? Which one is active? Can it be possible to check them all?

ksschwabe commented:
@magicrub 

The configuration of the USART is done in the hw_config.h file. I assume you are using the Pixhawk. Here is the extract of the file that sets all of the configuration parameters for the Pixhawk.
```
/****************************************************************************
 * TARGET_HW_PX4_FMU_V2
 ****************************************************************************/

#elif  defined(TARGET_HW_PX4_FMU_V2)

# define APP_LOAD_ADDRESS               0x08004000
# define BOOTLOADER_DELAY               5000
# define BOARD_FMUV2
# define INTERFACE_USB                  1
# define INTERFACE_USART                1
# define USBDEVICESTRING                "PX4 BL FMU v2.x"
# define USBPRODUCTID                   0x0011
# define BOOT_DELAY_ADDRESS             0x000001a0

# define BOARD_TYPE                     9
# define _FLASH_KBYTES                  (*(uint16_t *)0x1fff7a22)
# define BOARD_FLASH_SECTORS            ((_FLASH_KBYTES == 0x400) ? 11 : 23)
# define BOARD_FLASH_SIZE               (_FLASH_KBYTES * 1024)

# define OSC_FREQ                       24

# define BOARD_PIN_LED_ACTIVITY         0               // no activity LED
# define BOARD_PIN_LED_BOOTLOADER       GPIO12
# define BOARD_PORT_LEDS                GPIOE
# define BOARD_CLOCK_LEDS               RCC_AHB1ENR_IOPEEN
# define BOARD_LED_ON                   gpio_clear
# define BOARD_LED_OFF                  gpio_set

# define BOARD_USART  					USART2
# define BOARD_USART_CLOCK_REGISTER 	RCC_APB1ENR
# define BOARD_USART_CLOCK_BIT      	RCC_APB1ENR_USART2EN

# define BOARD_PORT_USART   			GPIOD
# define BOARD_PORT_USART_AF 			GPIO_AF7
# define BOARD_PIN_TX     				GPIO5
# define BOARD_PIN_RX		     		GPIO6
# define BOARD_USART_PIN_CLOCK_REGISTER RCC_AHB1ENR
# define BOARD_USART_PIN_CLOCK_BIT  	RCC_AHB1ENR_IOPDEN

# define BOARD_FORCE_BL_PIN_OUT         GPIO14
# define BOARD_FORCE_BL_PIN_IN          GPIO11
# define BOARD_FORCE_BL_PORT            GPIOE
# define BOARD_FORCE_BL_CLOCK_REGISTER  RCC_AHB1ENR
# define BOARD_FORCE_BL_CLOCK_BIT       RCC_AHB1ENR_IOPEEN
# define BOARD_FORCE_BL_PULL            GPIO_PUPD_PULLUP
```

Currently, the USART port is configured for USART2 on pins D5 and D6, which breakout on connector J202. (I don't have a Pixhawk, so I am just getting this from the schematics). If you want to change which USART port it uses, you can do that by just changing the configs in the hw_confgi.h file. (Remember to set the BOARD_PORT_USART_AF correctly - you can look it up in the STM32F427 datasheet).

Currently, the bootloader supports only a single USART port for bootloading.
